adsft/extraction.py

```diff
@@ -608,29 +608,29 @@ def parse_xml(self, preferred_parser_names=None):
 
         for parser_name in preferred_parser_names:
             parsed_xml = self._parse_xml(parser_name)
+
             logger.debug("Checking if the parser '{}' succeeded".format(parser_name))
             success = False
-            for xpath in META_CONTENT[self.meta_name].get('fulltext', {}).get('xpath', []):
-                fulltext = None
-                fulltext_elements = parsed_xml.xpath(xpath)
-                if len(fulltext_elements) > 0:
-                    fulltext = u" ".join(map(unicode.strip, map(unicode, fulltext_elements[0].itertext())))
-                    fulltext = TextCleaner(text=fulltext).run(decode=False, translate=True, normalise=True, trim=True)
-                if not fulltext:
-                    continue
-                else:
-                    logger.debug("The parser '{}' succeeded".format(parser_name))
-                    success = True
+            ft_found = False
+            for content_name in META_CONTENT[self.meta_name]:
+                if ft_found:
                     break
-
-            if not success:
-                logger.debug("The parser '{}' failed".format(parser_name))
-            else:
+                for xpath in META_CONTENT[self.meta_name].get(content_name, {}).get('xpath', []):
+                    element_found = False
+                    elements = parsed_xml.xpath(xpath)
+                    if len(elements) > 0:
+                        element_found = True
+                        success = True
+                    if content_name == 'fulltext' and element_found:
+                        ft_found = True
+                        fulltext = u" ".join(map(unicode.strip, map(unicode, elements[0].itertext())))
+                        fulltext = TextCleaner(text=fulltext).run(decode=False, translate=True, normalise=True, trim=True)
+                        break
+            if success:
+                logger.debug("The parser '{}' succeeded".format(parser_name))
                 break
-
-        if not success:
-            logger.warn('Parsing XML in non-standard way')
-            parsed_xml = lxml.html.document_fromstring(self.raw_xml.encode('utf-8'))
+            else:
+                logger.debug("The parser '{}' failed".format(parser_name))

```

Comment:
Instead of succeeded or failed, could we say something like this?

```
"The parser '{}' did not extract any of the following fields '{}'".format(parser_name, ", ".join(META_CONTENT[self.meta_name].keys()))
```
And:
```
"The parser '{}' succeeded extracting the following fields '{}'".format(parser_name, ", ".join(content_names))
```
Where `content_names` would be an array you can construct in the loop. But then, to be complete, we would need to remove these lines:
```
if ft_found:
    break
```
This would give us a quick first diagnostic instead of a generic success/fail message, and hopefully it might help us in our investigations.

Reply:
I'll add it. 

Just to note, this will change the logic so that we will loop through all parts of the fulltext instead of stopping when we find the body (which includes an empty body). This makes sense since we are no longer using the body to define success.